jit-compiler/src/codegen.rs

```diff
@@ -315,6 +336,88 @@ impl<'a, T: FieldElement> CodeGenerator<'a, T> {
     }
 }
 
+/// Used for patterns in match and let statements:
+/// `value_name` is an expression string that is to be matched against `pattern`.
+/// Returns a rust pattern string (tuple of new variables, might be nested) and a code string
+/// that, when executed, returns an Option with the values for the new variables if the pattern
+/// matched `value_name` and `None` otherwise.
+///
+/// So if `let (vars, code) = check_pattern("x", pattern)?;`, then the return value
+/// can be used like this: `if let Some({vars}) = ({code}) {{ .. }}`
+fn check_pattern(value_name: &str, pattern: &Pattern) -> Result<(String, String), String> {
+    Ok(match pattern {
+        Pattern::CatchAll(_) => ("()".to_string(), "Some(())".to_string()),
+        Pattern::Number(_, n) => (
+            "_".to_string(),
+            format!(
+                "({value_name}.clone() == {}).then_some(())",
+                format_signed_integer(n)
+            ),
+        ),
+        Pattern::String(_, s) => (
+            "_".to_string(),
+            format!("({value_name}.clone() == {}).then_some(())", quote(s)),
+        ),
+        Pattern::Tuple(_, items) => {
+            let mut vars = vec![];
+            let inner_code = items
+                .iter()
+                .enumerate()
+                .map(|(i, item)| {
+                    let (v, code) = check_pattern(&format!("{value_name}.{i}"), item)?;
+                    vars.push(v);
+                    Ok(format!("({code})?"))
+                })
+                .collect::<Result<Vec<_>, String>>()?
+                .join(", ");
+            (
+                format!("({})", vars.join(", ")),
+                format!("(|| Some(({inner_code})))()"),
+            )
+        }
+        Pattern::Array(_, items) => {
+            let mut vars = vec![];
+            let mut ellipsis_seen = false;
+            let inner_code = items
+                .iter()
+                .enumerate()
+                .filter_map(|(i, item)| {
+                    if matches!(item, Pattern::Ellipsis(_)) {
+                        ellipsis_seen = true;
+                        return None;
+                    }
+                    Some(if ellipsis_seen {
+                        let i_rev = items.len() - i;
+                        (format!("{value_name}[{value_name}.len() - {i_rev}]"), item)
+                    } else {
+                        (format!("{value_name}[{i}]"), item)
+                    })
+                })
+                .map(|(access_name, item)| {
```

Comment:
i wonder if this `map` being merged into the preceding `filter_map` makes it easier to read?

Reply:
The issue is that the first is a filter, i.e. returns Option and the second one returns Result. I think it's too complicate to combine both.